### PR TITLE
Eco 778/add logout api

### PR DESCRIPTION
**Pull request by OrenZak**
#### Main purpose:
Allowing logout from a user
#### Technical description:
- `Kin.logout()` API which handles the logout for the current user.
- Remove relevant cached info from all the repos

This PR based on #250

### CODE REVIEW

**Comment by OrenZak** [comment]
This PR based on #250, make sure to sync with the epic after the merge.

File: common/src/main/java/com/kin/ecosystem/common/ObservableData.java
Comment (YohayBar) [inline common/src/main/java/com/kin/ecosystem/common/ObservableData.java:121]: this is dangerous.  one observer may cancel to all other

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthLocalData.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/auth/AuthLocalData.java]: do we want to delete Device ID?

Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/auth/AuthLocalData.java]: nop

File: common/src/main/java/com/kin/ecosystem/common/ObservableData.java
Comment (OrenZak) [reply common/src/main/java/com/kin/ecosystem/common/ObservableData.java:121]: what do you suggest?

Comment (YohayBar) [reply common/src/main/java/com/kin/ecosystem/common/ObservableData.java:121]: why do you want to remove observers? you can disable and enable the observer maybe?

Comment (OrenZak) [reply common/src/main/java/com/kin/ecosystem/common/ObservableData.java:121]: well, as i see it now, we allowing the developer to add observers although the user is not logged in so we should not remove them.
We should remove only the`paymentObservers` because we shouldn't listen for payments if the account not logged in.

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: should you also call
```
		cachedAuthToken = null;
		jwt = null;
		localData.logout();
```

Comment (YohayBar) [reply core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: i.e  call ` internalLogout()`

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: why do wait for a response?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java]: should we also check in login if user_id changed? If it did change we can throw an error - enforcing developer to explicitly call login before. Or we can logout implicitly?

File: core/src/main/java/com/kin/ecosystem/core/data/offer/OfferRepository.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/offer/OfferRepository.java]: should we also clear cachedOfferList?

File: core/src/main/java/com/kin/ecosystem/core/data/settings/SettingsDataSourceLocal.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/settings/SettingsDataSourceLocal.java]: In the future, we might want to save additional information in setting related to single public address, maybe keep  IS_BACKED_UP_KEY and append publicAddress to it as the key?

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: no need,`local.setJwt(...)` overrides the relevant data.

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: otherwise, jwt can be deleted and the logout call will fail.

File: core/src/main/java/com/kin/ecosystem/core/data/offer/OfferRepository.java
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/offer/OfferRepository.java]: only `cachedOfferList` 🥇

File: core/src/main/java/com/kin/ecosystem/core/data/settings/SettingsDataSourceLocal.java
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/settings/SettingsDataSourceLocal.java]: 👍

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: i'll change that so we won't wait.

File: common/src/main/java/com/kin/ecosystem/common/ObservableData.java
Comment (YohayBar) [reply common/src/main/java/com/kin/ecosystem/common/ObservableData.java:121]: I'm not sure we need to remove observer - if we log out this observer should not omit any value

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: if throw new UserLoggedInException don't call localData.setJWT(jwtBody);
The error tells the Host app to first log out and only then login again

Comment (YohayBar) [reply core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: Also, we should call remote logout - we enforce the host app to do logout before login in with different user

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java:474]: you can also set paymentRegistration and balanceRegistration to null

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java]: maybe call it log out if in the future there will be an additional properties that need to be cleared

Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java]: it the future we will split if needed it doesn't matter now.

File: core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/auth/AuthRepository.java]: it's an internal exception not related to the DS.
so `Kin.java` will know there is already logged in user. and clear other cached data.

File: sdk/src/main/java/com/kin/ecosystem/Kin.java
Comment (YohayBar) [inline sdk/src/main/java/com/kin/ecosystem/Kin.java]: If setJWT through exception we  can't assume it succeeded in setting the JWT

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java]: You can be explicit about using ArrayList<String>

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java]: the string will end with STRING_WALLETS_DELIMITER

Comment (YohayBar) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java]: sb.deleteCharAt(sb.length() - 1)

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java
Comment (YohayBar) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java:57]: is this logic unit tested?

File: sdk/src/main/java/com/kin/ecosystem/Kin.java
Comment (YohayBar) [inline sdk/src/main/java/com/kin/ecosystem/Kin.java:287]: OfferReposetiry?

File: sdk/src/main/java/com/kin/ecosystem/main/presenter/EcosystemPresenter.java
Comment (YohayBar) [inline sdk/src/main/java/com/kin/ecosystem/main/presenter/EcosystemPresenter.java]: add log just in case

File: sdk/src/main/java/com/kin/ecosystem/settings/presenter/SettingsPresenter.java
Comment (YohayBar) [inline sdk/src/main/java/com/kin/ecosystem/settings/presenter/SettingsPresenter.java]: add log just in case

**Review by YohayBar — APPROVED** [review]
see my comments otherwise 👍

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceLocal.java:57]: yes

File: sdk/src/main/java/com/kin/ecosystem/main/presenter/EcosystemPresenter.java
Comment (OrenZak) [reply sdk/src/main/java/com/kin/ecosystem/main/presenter/EcosystemPresenter.java]: 👍

**Pull request merged**
